osutil/disks/disks.go

```diff
@@ -0,0 +1,362 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package disks
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/snapcore/snapd/osutil"
+)
+
+var (
+	// for mocking in tests
+	devBlockDir = "/sys/dev/block"
+
+	luksUUIDPatternRe = regexp.MustCompile(`(?m)CRYPT-LUKS2-([0-9a-f]{32})`)
```

Comment:
is this documented somewhere? or some code pointer to it?  unless as I ask below unsure why the (?m) and the submatch

Reply:
the `(?m)` was to prevent the regexp from also matching something where an attacker drive has a partition name like `CRYPT-LUKS2-$UUID-other-externalfs`, because with that name, the uuid file from sysfs then has:
```
CRYPT-LUKS2-$UUID-CRYPT-LUKS2-other-externalfs
```

This is because the uuid from sysfs is of the format:
```
CRYPT-LUKS2-$UUID-$NAME
```

where $NAME is the verbatim name of the mapper volume which is under control of the attacker, and UUID is the partition uuid of the physical encrypted partition.

Though, now that I have refactored the code to just delete the name from the uuid file we read from sysfs, we can probably safely drop the `(?m)` since the name is no longer in the input to the regexp. Previously my code dynamically added the name to the regexp before compiling, but I decided to change it to it's current implementation as that seemed safer/simpler.

Also, unfortunately no I cannot find any documentation of this format, but I can find a code pointer if you like.

Reply:
mm actually I mis-spoke the (?m) actually doesn't do that, I don't remember why I had that there, it should have been `^` for the start of the line

Reply:
yes, I was expecting "^pattern$"

Reply:
this is changed now